Sources/ServiceLifecycle/ServiceGroup.swift

```diff
@@ -310,6 +333,49 @@ public actor ServiceGroup: Sendable, Service {
                 "We did not create a graceful shutdown manager per service"
             )
 
+            var _unownedTaskGroupHandledCarefully = group
+            group.addTask {
+                // This is the task that listens to added services and starts them while the group is running
+
+                await withTaskCancellationHandler {
+                    // Channel will be finished in `shutdownGracefully`, we must not add services after graceful shutdown has started
+                    for await serviceConfiguration in addedServiceChannel {
+                        self.logger.debug(
+                            "Starting added service",
+                            metadata: [
+                                self.loggingConfiguration.keys.serviceKey: "\(serviceConfiguration.service)"
+                            ]
+                        )
+
+                        let gracefulShutdownManager = GracefulShutdownManager()
+                        gracefulShutdownManagers.append(gracefulShutdownManager)
+                        services.append(serviceConfiguration)
+
+                        precondition(
+                            services.count == gracefulShutdownManagers.count,
+                            "Mismatch between services and graceful shutdown managers"
+                        )
+
+                        _unownedTaskGroupHandledCarefully.addServiceTask(
+                            serviceConfiguration,
+                            gracefulShutdownManager: gracefulShutdownManager,
+                            index: services.count - 1
+                        )
+                    }
+                } onCancel: {
+                    addedServiceChannel.finish()
+                }
```

Comment:
This shouldn't be necessary since the channel setups up cancellation handlers as well and the for loop above will finish.

Reply:
If that's the case it sounds like a bug in the `AsyncChannel` to me